### PR TITLE
Reduce the number of segments to wait for convergence when rebalancing

### PR DESCRIPTION
**Pull request by saurabhd336**
Instructions:
1. The PR has to be tagged with at least one of the following labels (*):
   1. `feature`
   2. `bugfix`
   3. `performance`
   4. `ui`
   5. `backward-incompat`
   6. `release-notes` (**)
2. Remove these instructions before publishing the PR.
 
(*) Other labels to consider:
- `testing`
- `dependencies`
- `docker`
- `kubernetes`
- `observability`
- `security`
- `code-style`
- `extension-point`
- `refactor`
- `cleanup`

(**) Use `release-notes` label for scenarios like:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java [hunk 1]
```diff
@@ -304,11 +304,18 @@ public RebalanceResult rebalance(TableConfig tableConfig, Configuration rebalanc
...
-      IdealState idealState;
+
+    do {
+      Map<String, Map<String, String>> nextAssignment =
+          getNextAssignment(currentAssignment, targetAssignment, minAvailableReplicas, enableStrictReplicaGroup);
+      LOGGER.info("Got the next assignment for table: {} with number of segments to be moved to each instance: {}",
+          tableNameWithType,
+          SegmentAssignmentUtils.getNumSegmentsToBeMovedPerInstance(currentAssignment, nextAssignment));
+
+      Set<String> segmentsToMove = SegmentAssignmentUtils.getSegmentsToMove(currentAssignment, nextAssignment);
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:315]: `segmentsToMove` can be calculated with current and target assignment. We can keep the current logic of calculating next assignment.

**Review by Jackie-Jiang — COMMENTED** [review]
Mostly good

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/SegmentAssignmentUtils.java [hunk 2]
```diff
@@ -272,6 +272,7 @@ public static Map<String, String> getInstanceStateMap(Collection<String> instanc
     return instanceStateMap;
   }
 
+
```
[the remark this answers is not in the document]
Comment (saurabhd336) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/SegmentAssignmentUtils.java:275]: Ack

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/SegmentAssignmentUtils.java [hunk 3]
```diff
@@ -292,6 +293,21 @@ public static Map<String, Integer> getNumSegmentsToBeMovedPerInstance(Map<String
...
 
+  public static Set<String> getSegmentsToMove(Map<String, Map<String, String>> oldAssignment,
+      Map<String, Map<String, String>> newAssignment) {
+    Set<String> result = new HashSet<>();
+    for (Map.Entry<String, Map<String, String>> entry : newAssignment.entrySet()) {
+      String segmentName = entry.getKey();
+      Set<String> newInstancesAssigned = entry.getValue().keySet();
+      Set<String> oldInstancesAssigned = oldAssignment.get(segmentName).keySet();
+      if (!newInstancesAssigned.containsAll(oldInstancesAssigned) || !oldInstancesAssigned.containsAll(
+          newInstancesAssigned)) {
```
[the remark this answers is not in the document]
Comment (saurabhd336) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/SegmentAssignmentUtils.java:304]: Ack

**Review by Jackie-Jiang — APPROVED** [review]

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**